pkg/infrastructure/nutanix/clusterapi/clusterapi.go

```diff
@@ -0,0 +1,212 @@
+	return nutanixtypes.Name
+}
+
+// PreProvision creates the resources required prior to running capi nutanix controller.
+func (p Provider) PreProvision(ctx context.Context, in infracapi.PreProvisionInput) error {
+	// create categories with name "kubernetes-io-cluster-<cluster_id>" and values ["owned", "shared"].
+	// load the rhcos image to prism_central.
+
+	ic := in.InstallConfig.Config
+	nutanixCl, err := nutanixtypes.CreateNutanixClientFromPlatform(ic.Platform.Nutanix)
+	if err != nil {
+		return fmt.Errorf("fail to create nutanix client. %w", err)
+	}
+
+	// create the category key.
+	categoryKey := nutanixtypes.CategoryKey(in.InfraID)
+	keyBody := &nutanixclientv3.CategoryKey{
+		APIVersion:  ptr.To("3.1.0"),
+		Description: ptr.To("Openshift Cluster Category Key"),
+		Name:        ptr.To(categoryKey),
+	}
+	respk, err := nutanixCl.V3.CreateOrUpdateCategoryKey(ctx, keyBody)
+	if err != nil {
+		return fmt.Errorf("failed to create the category key %q. %w", categoryKey, err)
+	}
+	logrus.Infof("Created the category key %q", *respk.Name)
+
+	// create the category value "owned".
+	valBody := &nutanixclientv3.CategoryValue{
+		Description: ptr.To("Openshift Cluster Category Value: resources owned by the cluster"),
+		Value:       ptr.To(nutanixtypes.CategoryValueOwned),
+	}
+	respv, err := nutanixCl.V3.CreateOrUpdateCategoryValue(ctx, categoryKey, valBody)
+	if err != nil {
+		return fmt.Errorf("failed to create the category value %q with category key %q. %w", nutanixtypes.CategoryValueOwned, categoryKey, err)
+	}
+	logrus.Infof("Created the category value %q with name %q", *respv.Value, *respv.Name)
+
+	// create the category value "shared".
+	valBody = &nutanixclientv3.CategoryValue{
+		Description: ptr.To("Openshift Cluster Category Value: resources used but not owned by the cluster"),
+		Value:       ptr.To(nutanixtypes.CategoryValueShared),
+	}
+	respv, err = nutanixCl.V3.CreateOrUpdateCategoryValue(ctx, categoryKey, valBody)
+	if err != nil {
+		return fmt.Errorf("failed to create the category value %q with category key %q. %w", nutanixtypes.CategoryValueShared, categoryKey, err)
+	}
+	logrus.Infof("Created the category value %q with name %q", *respv.Value, *respv.Name)
+
+	// upload the rhcos image.
+	imgName := nutanixtypes.RHCOSImageName(in.InfraID)
+	imgURI := string(*in.RhcosImage)
+	imgReq := &nutanixclientv3.ImageIntentInput{}
+	imgSpec := &nutanixclientv3.Image{
+		Name:        &imgName,
+		Description: ptr.To("Created By OpenShift Installer"),
+		Resources: &nutanixclientv3.ImageResources{
+			ImageType: ptr.To("DISK_IMAGE"),
+			SourceURI: &imgURI,
+		},
+	}
+	imgReq.Spec = imgSpec
+	imgMeta := &nutanixclientv3.Metadata{
+		Kind:       ptr.To("image"),
+		Categories: map[string]string{categoryKey: nutanixtypes.CategoryValueOwned},
+	}
+	imgReq.Metadata = imgMeta
+	respi, err := nutanixCl.V3.CreateImage(ctx, imgReq)
+	if err != nil {
+		return fmt.Errorf("failed to create the rhcos image %q. %w", imgName, err)
```

Comment:
nit: for fmt.Errorf you should use `:` instead of `.` before the `%w` verb

Reply:
Fixed.